File: demo/AWS_Demo_PHP/CreativeCloudAPI/FolderBrowse.php
<?php

class FolderBrowse
{
	private $sid;// It has been provided from Creative Cloud Web	
	private static $api = '/inforelay/browsefolder/';// Service api
	private $userid;// The user's Email address		
	private $pwd;// The user password would require using lower-case letters and encrypt the password to its MD5 hash value
	
	function __construct($sid, $userid, $pwd, $token, $inforelay)
	{		
		$this->sid = $sid; 
		$this->userid = $userid; 
		$this->pwd = $pwd; 
		$this->token = $token; 
		$this->inforelay = $inforelay; 
	} 
	
	/* Represents an entire HTML or XML document.
	 * Preparing for DOMDocument. This class is available at: 
	 * http://www.php.net/manual/en/class.domdocument.php
	 * */	
	function payload($folderID, $type, $pageno, $pagesize, $sortby, $sortdirection)
	{
		$page = '';// The tag name is "page"
		
		$dom = new DOMDocument('1.0');
		$dom->encoding = 'UTF-8';
		$root = $dom->createElement('browse');// The root element of request payload 
				$dom->appendChild($root);
		$item = $dom->createElement('token', $this->token);
				$root->appendChild($item);
		$item = $dom->createElement('language', 'zh_TW');// The service language
				$root->appendChild($item);
		$item = $dom->createElement('userid', $this->userid);
				$root->appendChild($item);
		$item = $dom->createElement('folderid', $folderID);
				$root->appendChild($item);
		if( !is_null($type))
		{
		$item = $dom->createElement('type', $type);
				$root->appendChild($item);
		}
		$itemPage = $dom->createElement('pageno', $pageno);
				$item->appendChild($itemPage);	
		$itemPage = $dom->createElement('pagesize', $pagesize);
				$item->appendChild($itemPage);		
		$item = $dom->createElement('sortby', $sortby);
				$root->appendChild($item);		
		$item = $dom->createElement('sortdirection', $sortdirection);
				$root->appendChild($item); 
		$xmlStr = $dom->saveXML();
	
		$this->xmlStr = $xmlStr; 
	}
	
	function getResponse()
	{			
		$dataCenter = $this->inforelay;// Connection server		
		$url = 'https://'.$dataCenter.self::$api;
		
		$header = array("Cookie:ONE_VER=1_0;sid=$this->sid;path=/");// Cookies have to add the value of SID
		
		// Initialise and execute a cURL request
		$ch = curl_init();
		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 30);
		curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
		curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HEADER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
		curl_setopt($ch, CURLOPT_POST, 1);// Doing a regular HTTP POST
		curl_setopt($ch, CURLOPT_POSTFIELDS, $this->xmlStr);
		
		// Get the response from the server		
		$response = curl_exec($ch);

		//HTTP error
		if(curl_errno($ch))
		{
			echo '<br>'.curl_error($ch);
			exit('<br>Error : You have to check HttpsURLConnection or Input payload.<br>');
		}
		
		curl_close($ch); //close curl
		
		/* Get the response from the server and parse it */	
		$pos = strpos($response, "<?xml version='1.0' encoding='utf-8'?>");		
		$output = substr($response, $pos);// Find out output payload
		
		return $this->parse($output);// Parsing XML payload
	}
	
	/* Get the response from the server and parse it */	
	function parse($xml) 
	{
		$xml = simplexml_load_string($xml);
				
		$FolderBrowseResponse = array();
		
		foreach($xml->children() as $child)
		{							
			$parentfolder = array();
			$page = array();
			$folder = array();
			$file = array();			
				
			if( $child->getName() == 'status')
			{
				$status = $child;
				$FolderBrowseResponse['status'] = $status;
			}	
			else if( $child->getName() == 'rawfoldername')
			{
				$rawfoldername = $child;
				$FolderBrowseResponse['rawfoldername'] = $rawfoldername;
			}	
			else if( $child->getName() == 'parent')
			{
				$parent = $child;
				$FolderBrowseResponse['parent'] = $parent;
			}	
			else if( $child->getName() == 'rootfolderid')
			{
				$rootfolderid = $child;
				$FolderBrowseResponse['rootfolderid'] = $rootfolderid;
			}				
			else if( $child->getName() == 'owner')
			{
				$owner = $child;
				$FolderBrowseResponse['owner'] = $owner;
			}
			
			switch($child->getName())
			{
				case 'page':	
					$page['page'] = 'page';
					foreach($child->children() as $pagechild)
					{
						switch($pagechild->getName())
						{
							case 'pageno':   		 	$page['pageno'] = $pagechild; break;
							case 'pagesize':   			$page['pagesize'] = $pagechild; break;	
							case 'totalcount':   		$page['totalcount'] = $pagechild; break;	
							case 'hasnextpage':   		$page['hasnextpage'] = $pagechild; break;	
						}
					}					
					array_push($FolderBrowseResponse, $page);
					break;
				case 'folder':
					$folder['folder'] = 'folder';				
					foreach($child->children() as $folderchild)
					{
						switch($folderchild->getName())
						{
							case 'id':   		 		$folder['id'] = $folderchild; break;
							case 'rawfoldername':   	$folder['rawfoldername'] = $folderchild; break;	
							case 'treesize':   			$folder['treesize'] = $folderchild; break;	
							case 'isgroupaware':   		$folder['isgroupaware'] = $folderchild; break;	
							case 'isbackup':   			$folder['isbackup'] = $folderchild; break;	
							case 'isorigdeleted':   	$folder['isorigdeleted'] = $folderchild; break;	
							case 'ispublic':   			$folder['ispublic'] = $folderchild; break;	
							case 'createdtime':   		$folder['createdtime'] = $folderchild; break;	
							case 'markid':   			$folder['markid'] = $folderchild; break;	
						}
					}
					array_push($FolderBrowseResponse, $folder);
					break;
				case 'file': 	
					$file['file'] = 'file';				
					foreach($child->children() as $filechild)
					{
						switch($filechild->getName())
						{
							case 'id':   		 		$file['id'] = $filechild; break;
							case 'rawfilename':   		$file['rawfilename'] = $filechild; break;	
							case 'isgroupaware':   		$file['isgroupaware'] = $filechild; break;	
							case 'size':   				$file['size'] = $filechild.' Byte'; break;	
							case 'isbackup':   			$file['isbackup'] = $filechild; break;	
							case 'isorigdeleted':   	$file['isorigdeleted'] = $filechild; break;	
							case 'isinfected':   		$file['isinfected'] = $filechild; break;	
							case 'ispublic':   			$file['ispublic'] = $filechild; break;	
							case 'headversion':   		$file['headversion'] = $filechild; break;	
							case 'createdtime':   		$file['createdtime'] = $filechild; break;	
							case 'markid':   			$file['markid'] = $filechild; break;
						}
					}
					array_push($FolderBrowseResponse, $file);
					break;
			}//end switch			
			
			unset($page);
			unset($folder);
			unset($file);
		}		
		
		return $FolderBrowseResponse;			
	}
}
?>

File: demo/AWS_Demo_Java/src/com/ecareme/FolderBrowse.java
package com.ecareme;

import java.io.IOException;
import java.net.URL;
import java.util.LinkedList;
import javax.net.ssl.HttpsURLConnection;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import asuswebstorage.user.info.userInfo;
import entity.AcquireTokenResponse;
import entity.FolderBrowseResponse;
import entity.PageRsp;
import entity.PublicFileBase;
import entity.PublicFolderBase;

public class FolderBrowse 
{	
	
	static String API = "/inforelay/browsefolder/";// InfoRelay API
	
	public static void main(String[] args) throws Exception
	{	
		FolderBrowse folderBrowse = new FolderBrowse();
		FolderBrowseResponse rsp = folderBrowse.getResponse();		
		System.out.println(rsp.toString());
	}

	public FolderBrowseResponse getResponse() throws Exception
	{
		/* Fetching token from RequestServiceGateway by userid & pwd. Then you will get token, Inforelay and Webrelay domain/IP */
		AcquireToken at = new AcquireToken();
		AcquireTokenResponse acquireTokenResponse = null;
		try {
			acquireTokenResponse = at.getResponse(userInfo.userid, userInfo.pwd);
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}

		// The correct AcquireToken response payload's status value must equals "0
		if (acquireTokenResponse == null || !"0".equals(acquireTokenResponse.getStatus())) {
			throw new Exception("Error : You have to check the AcquireToken api response");
		}
		
		String token = acquireTokenResponse.getToken();// Token value
		String server = acquireTokenResponse.getInfoRelay();// Connection server

		String urlstr = server + API;
		URL url = new URL(urlstr);

		/* The HttpsURLConnection start */
		HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
		connection.setConnectTimeout(60 * 1000); 
		connection.setReadTimeout(60 * 1000);

		StringBuilder cookie = new StringBuilder();
		cookie.append("sid=").append(userInfo.sid).append(";");// Cookies have to add the value of SID

		connection.addRequestProperty("cookie", cookie.toString());// User must add the cookies in the header
		connection.setRequestMethod("POST");
		connection.setDoOutput(true);
		connection.setDoInput(true);			
		try {
			connection.connect();
		} catch (IOException ioe) {
			System.err.println("Get Connection Error:" + ioe.getMessage());
			throw ioe;
		} 		 		

		/* Preparing for DocumentBuilderFactory. This class is available at: 
		 * http://download.oracle.com/javase/1.4.2/docs/api/javax/xml/parsers/DocumentBuilderFactory.html
		 * */		
		String root = "browse";// The root element of request payload  
		
		String language = "zh_TW";// The service language
		String folderid = "82824190";// Browsing the specify folder's ID
		//String page = "";// The tag name is "page"
		String pageno = "1";// Page number
		String pagesize = "10";// Amount of information access granted to a page 
		String sortby = "1"; // The way of sorting file: 1 = Sort the catalog by file(or folder) name | 2 = Sort the catalog by time 
		String sortdirection = "0";// The way of sorting file: 0 = ASC | 1 = DESC 

		String[] elmName = { "token", "language", "userid", "folderid",	"type", "pageno", "pagesize", "sortby", "sortdirection" };// Define each XML tag name			
		String[] data = { token, language, userInfo.userid, folderid, "", pageno, pagesize, sortby, sortdirection };// Set each value of tag
		
		/* To create XML documents. DocumentBuilderFactory can obtain a parser that produces DOM object trees from XML documents */
		DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory
				.newInstance();
		DocumentBuilder documentBuilder = documentBuilderFactory
				.newDocumentBuilder();
		Document document = documentBuilder.newDocument();
		Element rootElement = document.createElement(root);
		document.appendChild(rootElement);
		Element elm;
		int i;
		for (i = 0; i < data.length; i++) {
			elm = document.createElement(elmName[i]);
			elm.appendChild(document.createTextNode(data[i]));
			rootElement.appendChild(elm);
		}
		
		/* Used to process XML from a variety of sources and write the transformation output to server */
		TransformerFactory transformerFactory = TransformerFactory.newInstance();
		Transformer transformer = transformerFactory.newTransformer();
		DOMSource source = new DOMSource(document);
		StreamResult result = new StreamResult(connection.getOutputStream());
		transformer.transform(source, result);

		/* Get the response from the server and parse it */						
		DocumentBuilderFactory documentBuilderFactoryResponse = DocumentBuilderFactory.newInstance();
		DocumentBuilder documentBuilderResponse = documentBuilderFactoryResponse.newDocumentBuilder();
		Document documentResponse = documentBuilderResponse.parse(connection.getInputStream());
		Element rootResponse = (Element) documentResponse.getDocumentElement();
		NodeList nodelist;
		nodelist = rootResponse.getChildNodes();

		FolderBrowseResponse rsp = new FolderBrowseResponse();		
		PageRsp pagersp = new PageRsp();				
		LinkedList<PublicFolderBase> pfolder =new LinkedList<PublicFolderBase>(); 
		LinkedList<PublicFileBase> pfile = new LinkedList<PublicFileBase>();
		
		// Compose the payload
		int j, k, m;		
		for (j = 0; j < nodelist.getLength(); j++) {
			Node node = nodelist.item(j);
			
			/* status */
			if(node.getNodeName().equals("status")){
				rsp.setStatus(node.getTextContent());
			}	
			
			/*rawfoldername*/
			if (node.getNodeName().equals("rawfoldername")) {
				rsp.setRawFolderName(node.getTextContent());
			}
			
			/* parent*/			
			if(node.getNodeName().equals("parent")){	
				rsp.setParent(node.getTextContent());
			}
			
			/* root folder id*/
			if (node.getNodeName().equals("rootfolderid")) {
				rsp.setRootFolderID(node.getTextContent());
			}
			
			/* page */	
			if(node.getNodeName().equals("page")){	
				for(k = 0; k < node.getChildNodes().getLength(); k++){
					if(node.getChildNodes().item(k).getNodeName().equals("pageno")){
						pagersp.setPageno(Integer.parseInt(node.getChildNodes().item(k).getTextContent()));			
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("pagesize")){
						pagersp.setPagesize(Integer.parseInt(node.getChildNodes().item(k).getTextContent()));	
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("totalcount")){
						pagersp.setTotalcount(Integer.parseInt(node.getChildNodes().item(k).getTextContent()));	
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("hasnextpage")){
						pagersp.setHasnextpage(Integer.parseInt(node.getChildNodes().item(k).getTextContent()));	
					}
				}				
			}

			/* folder */
			if(node.getNodeName().equals("folder")){	
				PublicFolderBase publicfolderbase = new PublicFolderBase();
				for(k = 0; k < node.getChildNodes().getLength(); k++){							
					if(node.getChildNodes().item(k).getNodeName().equals("id")){
						publicfolderbase.setId(node.getChildNodes().item(k).getTextContent());			
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("rawfoldername")){
						publicfolderbase.setRawFolderName(node.getChildNodes().item(k).getTextContent());	
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("treesize")){
						publicfolderbase.setTreeSize(node.getChildNodes().item(k).getTextContent());	
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("isgroupaware")){	
						boolean b =new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfolderbase.setIsGroupAware(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("isbackup")){	
						boolean b =new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfolderbase.setIsbackup(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("isorigdeleted")){	
						boolean b =new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfolderbase.setIsorigdeleted(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("ispublic")){	
						boolean b =new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfolderbase.setIspublic(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("createdtime")){
						publicfolderbase.setCreatedtime(node.getChildNodes().item(k).getTextContent());	
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("markid")){
						publicfolderbase.setMarkid(node.getChildNodes().item(k).getTextContent());							
					}
				}				
				pfolder.add(publicfolderbase);	
			}			

			/* file */	
			if(node.getNodeName().equals("file")){
				PublicFileBase publicfilebase = new PublicFileBase();
				for(k = 0; k < node.getChildNodes().getLength(); k++){
					if(node.getChildNodes().item(k).getNodeName().equals("id")){
						publicfilebase.setId(node.getChildNodes().item(k).getTextContent());
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("rawfilename")){
						publicfilebase.setRawFileName(node.getChildNodes().item(k).getTextContent());
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("size")){
						publicfilebase.setSize(Long.parseLong(node.getChildNodes().item(k).getTextContent()));
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("isgroupaware")){
						boolean b=new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfilebase.setIsGroupAware(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("isbackup")){
						boolean b=new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfilebase.setIsbackup(b);
					}				
					else if(node.getChildNodes().item(k).getNodeName().equals("isorigdeleted")){
						boolean b=new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfilebase.setIsorigdeleted(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("isinfected")){
						publicfilebase.setIsinfected(Integer.parseInt(node.getChildNodes().item(k).getTextContent()));
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("ispublic")){
						boolean b=new Boolean(node.getChildNodes().item(k).getTextContent()).booleanValue();
						publicfilebase.setIspublic(b);
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("headversion")){
						publicfilebase.setHeadversion(node.getChildNodes().item(k).getTextContent());
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("createdtime")){
						publicfilebase.setCreatedtime(node.getChildNodes().item(k).getTextContent());
					}
					else if(node.getChildNodes().item(k).getNodeName().equals("markid")){
						publicfilebase.setMarkid(node.getChildNodes().item(k).getTextContent());
					}
				}
				pfile.add(publicfilebase);					
			}
		}
		rsp.setPage(pagersp);	
		rsp.setFolders(pfolder);
		rsp.setFiles(pfile);

		// The correct response payload's status value must equals "0"
		if ( !"0".equals(rsp.getStatus()) )
		{
			System.out.println("status:"+rsp.getStatus());
			throw new Exception("Error : You have to check FolderBrowse api response");
		}
		
		return rsp;
	}
}
